### DISCUSSION TITLE
How to install `require-dev` dependencies together with `composer require package-name`?

### DISCUSSION THREAD

**Discussion by MurzNN** [composer/composer#11451] [Q&A / Support] [answered]
I need to install packages from the list of `require-dev` dependencies when installing a package using the command:
```
composer require package-name
```
But it installs only packages, listed in the `require` section, and skips all in `require-dev`.

And I can't find any key to enable this in the documentation. 

So, is there any way to automate installing of  `require-dev` dependencies via composer, together with a package?

**Comment by Seldaek** [comment]
Do you have COMPOSER_NO_DEV env var set? Otherwise by default it should always install require-dev.

Oh or maybe it's a misunderstanding.. Composer installs your project's require-dev always, but never your dependencies require-dev section, as those are only useful when developing the package in question, not when using the package as a dependency. That's the whole point of splitting things up in require-dev.

**Reply by MurzNN** [reply]
> but never your dependencies require-dev section

I need exactly this, to run unit tests in a dependent project. They can't be run via simply installing the dependent project separately, because it requires the framework core files.

Here is my project: https://git.drupalcode.org/project/opentelemetry and it requires `drupal/test_helpers` to make tests run successfully, and needs to be installed as a dependency to `drupal/core` project.

Yes, I can do it manually like this:
```
composer require drupal/opentelemetry
composer require --dev drupal/test_helpers
```

But it'll be much better to have something like this:
```
composer require drupal/opentelemetry --with-dev-dependencies
```

**Reply by MurzNN** [reply]
In Drupal.org pipeline it's seems workarounded via manually parsing the `require-dev` section and installing the listed packages via separate commands:
https://dispatcher.drupalci.org/job/drupal_contrib/686443/console
![image](https://user-images.githubusercontent.com/336662/236386045-2d94ed2e-c559-42a1-a9f6-d4e781982b34.png)

Here is the code that does this: https://git.drupalcode.org/project/drupalci_testbot/-/blob/dev/src/DrupalCI/Plugin/BuildTask/BuildStep/CodebaseAssemble/ComposerContrib.php#L100-105

**Reply by Seldaek** [reply, marked as answer]
Yeah I guess this is a side-effect of Drupal's historical baggage and the fact that (AFAIK) it cannot easily run from the vendor dir, so you need to install your project inside a drupal project to be able to test it, etc.. More modern projects typically do not have this limitation and then everything works more easily. 

Anyway there is no way we can allow this IMO as it would lead to way too many conflicts.

**Reply by kopeboy** [reply]
How can adding an additional option lead to conflicts for those not using the new option? 🤔

**Reply by stof** [reply]
@kopeboy it is not about those not using the option. It is about being unusable for people using that option.

including dev requirements of all dependents in the dependency resolution means Composer has to find a set of dependencies compatible with all those dev requirements of all packages at once. The chance of finding such a solution without dependency conflicts is very low. This description would be a direct ticket for the dependency hell.

**Comment by joachim-n** [comment]
I don't think this is a just a Drupal issue.
It was my understanding that the right way to develop and test a library package was to require it from a path repository that's a git clone. With that setup, you don't get the dev requirements of that package either.

**Reply by Seldaek** [reply]
@joachim-n No sure, but with that set up you can easily go in the path repo's dir, and run composer install to run tests in there etc.
